sys var: add `tidb_enable_shared_lock_upgrade` (#18447)
pingcap/docs: pessimistic-transaction.md

2. TiDB does not support `SELECT LOCK IN SHARE MODE`.

    TiDB does not support `SELECT LOCK IN SHARE MODE` by default.

>     You can enable [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) to be compatible with the `SELECT LOCK IN SHARE MODE` statement. The effect of the statement is the same as when there is no lock, and it will not block the read and write operations of other transactions.

Author: ```suggestion
    You can enable [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) to make TiDB compatible with the `SELECT LOCK IN SHARE MODE` statement. The effect of the statement is the same as when there is no lock, and it will not block the read and write operations of other transactions.
```

>     TiDB does not support `SELECT LOCK IN SHARE MODE` by default.
> 
>     You can enable [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) to make TiDB compatible with the `SELECT LOCK IN SHARE MODE` statement. The effect of the statement is the same as when there is no lock, and it will not block the read and write operations of other transactions.

Author: ```suggestion
    TiDB does not support `SELECT LOCK IN SHARE MODE` by default. You can enable [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) to make TiDB compatible with the `SELECT LOCK IN SHARE MODE` statement. The effect of the statement is the same as when there is no lock, and it will not block the read and write operations of other transactions.
```

>     Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_upgrade`](/system-variables.md#tidb_enable_shared_lock_upgrade-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement to add locks. However, note that the lock added at this time is not a true shared lock, but consistent with `SELECT FOR UPDATE`, which actually adds an exclusive lock. You can enable this variable when you want to be compatible with the `SELECT LOCK IN SHARE MODE` syntax while blocking writes to prevent data from being modified by parallel write transactions during reads. This variable takes effect regardless of the configuration of [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40).

Author: ```suggestion
    Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_promotion`](/system-variables.md#tidb_enable_shared_lock_upgrade-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement to add locks. However, note that the lock added at this time is not a true shared lock, but consistent with `SELECT FOR UPDATE`, which actually adds an exclusive lock. You can enable this variable when you want to be compatible with the `SELECT LOCK IN SHARE MODE` syntax while blocking writes to prevent data from being modified by parallel write transactions during reads. This variable takes effect regardless of the configuration of [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40).
```

2. TiDB does not support `SELECT LOCK IN SHARE MODE`.

    TiDB does not support `SELECT LOCK IN SHARE MODE` by default.

>     Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_upgrade`](/system-variables.md#tidb_enable_shared_lock_upgrade-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement. After enabling, the `SELECT LOCK IN SHARE MODE` statement will be upgraded to the `SELECT FOR UPDATE` statement, which adds pessimistic locks to the data rows.

Author: ```suggestion
    You can enable [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) to be compatible with the `SELECT LOCK IN SHARE MODE` statement. At this time, the effect of this statement is the same as not locking, and it will not block the read and write operations of other transactions.
```

>     Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_upgrade`](/system-variables.md#tidb_enable_shared_lock_upgrade-new-in-v8.3.0) system variable to enable the `SELECT LOCK IN SHARE MODE` statement. After enabling, the `SELECT LOCK IN SHARE MODE` statement will be upgraded to the `SELECT FOR UPDATE` statement, which adds pessimistic locks to the data rows.

Author: ```suggestion
    Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_upgrade`](/system-variables.md#tidb_enable_shared_lock_upgrade-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement. After enabling, the `SELECT LOCK IN SHARE MODE` statement will be upgraded to the `SELECT FOR UPDATE` statement, which adds pessimistic locks to the data rows.
```

>     Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_promotion`](/system-variables.md#tidb_enable_shared_lock_promotion-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement to add locks. However, note that the lock added at this time is not a true shared lock, but an exclusive lock, which is consistent with `SELECT FOR UPDATE`. If you want to block writes to prevent data from being modified by write transactions in parallel during reads while keeping TiDB compatible with the `SELECT LOCK IN SHARE MODE` syntax, you can enable this variable. This variable takes effect regardless of the configuration of [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40).

Author: ```suggestion
    Starting from v8.3.0, TiDB supports using the [`tidb_enable_shared_lock_promotion`](/system-variables.md#tidb_enable_shared_lock_promotion-new-in-v830) system variable to enable the `SELECT LOCK IN SHARE MODE` statement to add locks. However, note that the lock added at this time is not a true shared lock, but an exclusive lock, which is consistent with `SELECT FOR UPDATE`. If you want to block writes to prevent data from being modified by write transactions in parallel during reads while keeping TiDB compatible with the `SELECT LOCK IN SHARE MODE` syntax, you can enable this variable. Enabling this variable takes effect on the `SELECT LOCK IN SHARE MODE` statement, regardless of whether [`tidb_enable_noop_functions`](/system-variables.md#tidb_enable_noop_functions-new-in-v40) is enabled or not.
```
